File: leetcode/leetcode_p94/code_94.py
# 力扣第94题，二叉树的中序遍历
"""
给定一个二叉树，返回它的中序 遍历。
"""
from collections import deque

class TreeNode:
    def __init__(self, x) -> None:
        self.val = x
        self.left = None
        self.right = None

class Solution_00:
    def inorderTraversal(self, root):
        result = []

        def dfs(root):
            if not root:
                return []
            if isinstance(root.left, TreeNode):
                dfs(root.left)
            elif root.left:
                result.append(root.left)

            result.append(root.val)

            if isinstance(root.right, TreeNode):
                dfs(root.right)
            elif root.right:
                result.append(root.right)
        
        dfs(root)

        return result


class Solution_01:
    def inorderTraversal(self, root):
        if not root:
            return []
        result = []
        dq = deque()
        dq.append(root)

        while dq:
            cur = dq.popleft()
            if not isinstance(cur, TreeNode):
                result.append(cur)
                continue
            if cur.right:
                dq.appendleft(cur.right)
            
            dq.appendleft(cur.val)

            if isinstance(cur.left, TreeNode):
                dq.appendleft(cur.left)
            elif cur.left:
                result.append(cur.left)

        return result


if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.left.left = 3
    root.left.right = TreeNode(4)
    root.left.right.left = 5
    root.right = TreeNode(6)
    root.right.left = 7
    root.right.right = 8

    sl0 = Solution_00()
    result0 = sl0.inorderTraversal(root)
    print(result0)

    sl1 = Solution_01()
    result1 = sl1.inorderTraversal(root)
    print('result1:', result1)

File: leetcode/leetcode_p94/code_94.cpp
// 力扣第94题，二叉树的中序遍历
/*
给定一个二叉树，返回它的中序 遍历。
*/
#include <iostream>
#include <vector>
#include <algorithm>
#include <deque>
#include <stack>
using namespace std;


void print(vector<int> v);

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(NULL), right(NULL) {}
};


class Solution_00{
public:
    vector<int> result;
    void dfs(TreeNode* root){
        if (root->left){
            dfs(root->left);
        }
        
        result.emplace_back(root->val);

        if (root->right){
            dfs(root->right);
        }
    }
    vector<int> inorderTraversal(TreeNode* root) {
        if (!root){
            return {};
        }
        dfs(root);
        return result;
    }
};


class Solution_01{
public:
    vector<int> inorderTraversal(TreeNode* root) {
        if (!root)
            return {};
        deque<TreeNode*> dq;
        vector<int> result;

        dq.emplace_back(root);

        while (!dq.empty()){
            TreeNode* cur = dq.front();
            dq.pop_front();
            if (!cur->left && !cur->right){
               result.emplace_back(cur->val); 
               continue;
            } 

            if (cur->right)
                dq.emplace_front(cur->right);
            dq.emplace_front(new TreeNode(cur->val));

            if (cur->left)
                dq.emplace_front(cur->left);
        }
        return result;
    }
};


/*
官方方法二：栈
*/
class Solution_02{
public:
    vector<int> inorderTraversal(TreeNode* root){
        vector<int> res;
        stack<TreeNode*> stk;
        while (root != nullptr || !stk.empty()){
            while (root != nullptr){
                stk.push(root);
                root = root->left;
            }
            root = stk.top();
            stk.pop();
            res.push_back(root->val);
            root = root->right;
        }
        return res;
    }
};


int main()
{
    TreeNode tn(1);
    // tn.right = &(TreeNode(2));
    // tn.right->left = &TreeNode(3);

    Solution_00 sl0;
    vector<int> result0 = sl0.inorderTraversal(&tn);
    cout << "result0:\n";
    print(result0);

    cin.get();
    return 0;
}



void print(vector<int> v){
    for (int num: v)
        cout << num << ' ';
    cout << endl;
}